include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
+              }
+            guess_weights /= guess_weights_sum;
+            Assert(weights_are_ok(guess_weights), ExcInternalError());
+
+            // Gradient descent:
+            for (unsigned int outer_n = 0; outer_n < 10; ++outer_n)
+              {
+                const unsigned int dependent_direction = GeometryInfo<structdim>::vertices_per_cell - 1;
+                Tensor<1, GeometryInfo<structdim>::vertices_per_cell> current_gradient;
+                for (unsigned int row_n = 0;
+                     row_n < GeometryInfo<structdim>::vertices_per_cell;
+                     ++row_n)
+                  {
+                    if (row_n != dependent_direction)
+                      {
+                        current_gradient[row_n] = gradient_entry<spacedim, structdim>
+                                                  (row_n,
+                                                   dependent_direction,
+                                                   trial_point,
+                                                   guess_weights,
+                                                   step_size,
+                                                   get_point_from_weights);
+
+                        current_gradient[dependent_direction] -= current_gradient[row_n];
+                      }
+                  }
+
+                // since we need to travel in the -gradient direction, guess that
+                // we will go -0.5*gradient and do quasi-Newton iteration to pick
+                // the best multiplier:
+                double gradient_weight = -0.5;
+                auto gradient_weight_objective_function = [&](const double gradient_weight_guess)
+                                                          -> double
+                {
+                  return (trial_point -
+                  get_point_from_weights(guess_weights +
+                  gradient_weight_guess*current_gradient)).norm_square();
+                };
+
+                for (unsigned int inner_n = 0; inner_n < 10; ++inner_n)
+                  {
+                    const double update_numerator = centered_first_difference
+                                                    (gradient_weight, step_size, gradient_weight_objective_function);
+                    const double update_denominator = centered_second_difference
+                                                      (gradient_weight, step_size, gradient_weight_objective_function);
+
+                    // avoid division by zero. Note that we limit the gradient weight below
+                    if (std::abs(update_denominator) == 0.0)
+                      break;
+                    gradient_weight -= update_numerator/update_denominator;
+
+                    // Put a fairly lenient bound on the largest possible
+                    // gradient (things tend to be locally flat, so the gradient
+                    // itself is usually small)
+                    if (std::abs(gradient_weight) > 10)
```

Comment:
Do I see this right that the gradient_weight is a quantity with units `[length in reference cell]`? In that case, an absolute number makes sense here; otherwise, the comparison needs to be appropriately scaled.

Reply:
This is a dimensionless parameter: I apply a factor on the length of the gradient here to prevent the step from being too large. The gradient itself has the right units, but this is dimensionless.

Reply:
Thanks for confirming.